daffodil-core/src/main/scala/org/apache/daffodil/dsom/ChoiceGroup.scala

```diff
@@ -110,9 +110,12 @@ abstract class ChoiceTermBase(
   with ChoiceAGMixin
   with HasOptRepTypeMixinImpl {
 
-  requiredEvaluations(branchesAreNonOptional)
-  requiredEvaluations(branchesAreNotIVCElements)
-  requiredEvaluations(modelGroupRuntimeData.preSerialization)
+  override protected def check() {
+    super.check()
+    requiredEvaluations(branchesAreNonOptional)
+    requiredEvaluations(branchesAreNotIVCElements)
+    requiredEvaluations(modelGroupRuntimeData.preSerialization)
+  }
```

Comment:
This will need some scaladoc and a design note. 
A significant' change here is that in the past there were these factories. They did little except postpone the creation of the "real objects" until they were known participants in the schema. 

Upon  creation, the requiredEvaluation functions are registered, and *will* be evaluated later.

Now, we allow creation of the objects, and those objects which are always part of the "active schema"  like the schema set and processor factory object always can issue requiredEvaluations. But many other objects are created but we don't want to register their requiredEvaluations until it is known they are "in use by" the schema. Those have their calls to requiredEvaluations enclosed in a call to check(). Finally, a third class of things are only created on demand when another object, that has been determined to be "in use" causes them to be constructed. Those also require no qualification around their calls to requiredEvaluations(...) once constructed it is known they are used.

Reply:
Another comment addresses this also. It's because objects only want these requiredEcaluations to actually evaluate if the objects are "in use" by the schema. "in use" now isn't just "got constructed".

Reply:
Makes a lot of sense. Definitely needs some documntation. Thoughts on renaming the "check" function to something like ``registerAsActive`` to make it more clear what the purpose of it?